ecommerce/api.py

```diff
@@ -103,3 +111,184 @@ def make_reference_id(order):
+
+    Args:
+        coupons (QuerySet of coupon ids): List of coupons to filter for validity
+        auto_only (bool): Whether or not to filter by automatic=True
+
+    Returns:
+        list of CouponVersion: CouponVersion objects sorted by discount, highest first.
+    """
+    valid_coupons = []
+    now = now_in_utc()
+
+    # Get the ids of the latest coupon versions
+    cv_latest = (
+        CouponVersion.objects.select_related()
+        .filter(coupon__in=coupons)
+        .order_by("coupon", "-created_on")
+        .distinct("coupon")
+        .values_list("pk")
+    )
+
+    # filter by expiration and activation dates
+    query = (
+        CouponVersion.objects.select_related()
+        .filter(pk__in=cv_latest)
+        .filter(
+            Q(invoice_version__expiration_date__gte=now)
+            | Q(invoice_version__expiration_date__isnull=True)
+        )
+        .filter(
+            Q(invoice_version__activation_date__lte=now)
+            | Q(invoice_version__activation_date__isnull=True)
+        )
+    )
+
+    if auto_only:
+        query = query.filter(invoice_version__automatic=True)
+
+    # filter by redemption counts
+    for cv in query:
+        redemptions_global = CouponRedemption.objects.filter(coupon_version=cv)
+        redemptions_user = redemptions_global.filter(order__purchaser=user)
+        if (
+            cv.invoice_version.max_redemptions > redemptions_global.count()
+            and cv.invoice_version.max_redemptions_per_user > redemptions_user.count()
+        ):
+            valid_coupons.append(cv)
+    return sorted(valid_coupons, key=lambda x: x.invoice_version.amount, reverse=True)
+
+
+def get_eligible_coupons(product, code=None):
+    """
+    Return the eligible coupon ids for a product.  This may include both valid and invalid coupons.
+    Invalid coupons should be filtered out with the get_valid_coupon_versions function.
+
+    Args:
+        product (Product): product to filter CouponEligibility by
+        code (str): A coupon code to filter by
+
+    Returns:
+        TimestampedModelQuerySet: list of coupon ids that can be used with the products.
+    """
+    query = CouponEligibility.objects.select_related("coupon").filter(
+        product=product, coupon__enabled=True
+    )
+    if code:
+        query = query.filter(coupon__coupon_code=code)
+
+    return query.values_list("coupon", flat=True)
+
+
+def best_coupon_for_basket(basket, auto_only=False, code=None):
+    """
+    Get the best eligible coupon for the basket.
+    Assumes that the basket only contains one item/product.
+
+    Args:
+        basket (Basket): the basket Object
+        auto_only (bool): Only retrieve `automatic` Coupons
+        code (str): A coupon code to filter by
+
+    Returns:
+        CouponVersion: the CouponVersion with the highest discount, or None
+    """
+    basket_item = basket.basketitems.first()
```

Comment:
Are we validating that there is only one item in the basket before the function is called?

Reply:
I was thinking that would happen when creating/updating basket items?

Reply:
:+1: